### PR TITLE
TMDM-11044 [Impact Analysis] Adapt Low/Medium Impact to Foreign Key field

### PR DESCRIPTION
**Pull request by zfkang-talend**
Jira: https://jira.talendforge.org/browse/TMDM-11044

**What is the current behavior?** (You should also link to an open issue here)
If the FK field renamed, system will remove it and add new column, if this field contains data, the data would be lost.

**What is the new behavior?**
when renamed FK field, first query data, if this field contains data, the impact result of remove field would be High and create table, else, it would be Medium.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 1]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+        return fetchFieldCountByUserQuery(qb);
+    }
+
+    protected int fetchFieldCountByUserQuery(UserQueryBuilder qb) {
+        int count = 0;
+        try {
+            storage.begin();
+            StorageResults results = storage.fetch(qb.getSelect());
+            count = results.getCount();
+            results.close();
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: Should we move 'results.close()' into finally?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 2]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
             }
         }
+
+        Map<RemoveChange, ComplexTypeMetadata> renamedReferenceFieldMap = isFKRenamed(diffResult);
+        for (Map.Entry<RemoveChange, ComplexTypeMetadata> entry : renamedReferenceFieldMap.entrySet()) {
+            int count = fetchFieldCountOfNotNull(entry.getValue(), (FieldMetadata) entry.getKey().getElement());
+            if (count == 0) {
+                entry.getKey().setContainsData(false);
+            } else {
+                entry.getKey().setContainsData(true);
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: I would suggest to name 'entry.getKey()' as a local variable.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 3]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+            }
+        }
         return super.analyzeImpacts(diffResult);
     }
+
+    protected Map<RemoveChange, ComplexTypeMetadata> isFKRenamed(Compare.DiffResults diffResult) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: This function returns a map not boolean, but the function name starts with 'isXXXX', could it be changed to a more reasonable name?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 4]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+        UserQueryBuilder qb = UserQueryBuilder.from(entry).select(count()).where(not(emptyOrNull(field)));
+        return fetchFieldCountByUserQuery(qb);
+    }
+
+    protected int fetchFieldCountOfNull(ComplexTypeMetadata entry, FieldMetadata field) {
+        UserQueryBuilder qb = UserQueryBuilder.from(entry).select(count()).where(emptyOrNull(field));
+        return fetchFieldCountByUserQuery(qb);
+    }
+
+    protected int fetchFieldCountByUserQuery(UserQueryBuilder qb) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: For the above four functions, they use 'protected', could they just use 'private' access?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 5]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+        }
+        for (AddChange removeAction : diffResult.getAddChanges()) {
+            MetadataVisitable element = removeAction.getElement();
+            if (element instanceof ReferenceFieldMetadata) {
+                if (removeReferenceFieldMap.containsKey(((ReferenceFieldMetadata) element).getReferencedField())) {
+                    renamedReferenceFieldMap.put((removeReferenceFieldMap.get(((ReferenceFieldMetadata) element).getReferencedField())),
+                            ((ReferenceFieldMetadata) element).getContainingType().getEntity());
+                }
+            }
+        }
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: Above two for loop, if we may consider using concurrent thread to do it, like CountDownLatch  or Cyclicbarrier.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 6]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+    protected Map<RemoveChange, ComplexTypeMetadata> isFKRenamed(Compare.DiffResults diffResult) {
+        Map<FieldMetadata, RemoveChange> removeReferenceFieldMap = new HashMap<>();
+        Map<RemoveChange, ComplexTypeMetadata> renamedReferenceFieldMap = new HashMap<>();
+        for (RemoveChange removeAction : diffResult.getRemoveChanges()) {
+            MetadataVisitable element = removeAction.getElement();
+            if (element instanceof ReferenceFieldMetadata) {
+                removeReferenceFieldMap.put(((ReferenceFieldMetadata) element).getReferencedField(), removeAction);
+            }
+        }
+        for (AddChange removeAction : diffResult.getAddChanges()) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: It is not 'removeAction' here, should it be 'addChange' or 'addAction'?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 7]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+            MetadataVisitable element = removeAction.getElement();
+            if (element instanceof ReferenceFieldMetadata) {
+                if (removeReferenceFieldMap.containsKey(((ReferenceFieldMetadata) element).getReferencedField())) {
+                    renamedReferenceFieldMap.put((removeReferenceFieldMap.get(((ReferenceFieldMetadata) element).getReferencedField())),
+                            ((ReferenceFieldMetadata) element).getContainingType().getEntity());
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: Is it possible to change '((ReferenceFieldMetadata) element)' to a local variable?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java [hunk 8]
```diff
@@ -47,26 +53,72 @@ public HibernateStorageDataAnaylzer(HibernateStorage storage) {
...
+
+    protected int fetchFieldCountByUserQuery(UserQueryBuilder qb) {
+        int count = 0;
+        try {
+            storage.begin();
+            StorageResults results = storage.fetch(qb.getSelect());
+            count = results.getCount();
+            results.close();
+        } catch (Exception e) {
+            LOGGER.error("Hibernate Storage query data anaylzer failure", e);
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: Hibernate storage query data analyzer failure.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java (hunk 5 above)
Comment (lpw82gf) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/HibernateStorageDataAnaylzer.java]: We can go with single thread for now.

**Pull request merged**